kubeadm: allow creating a cluster with ECDSA keys

rojkov opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
Allow creating a cluster with ECDSA keys. The selected key type is defined in `InitConfiguration`'s `publicKeyAlgorithm` field. The default value for the option is `RSA` still.
Upon renewal the new certs are generated with the same algo used for the corresponding old certs.

**Which issue(s) this PR fixes**:
Contributes to kubernetes/kubeadm#1535

**Special notes for your reviewer**:
I decided to leave a new command line option for specifying encryption key algo out of scope of this PR since the flag would be removed anyway. The option is set via `InitConfiguration` only.

**Does this PR introduce a user-facing change?**:
```release-note
kubeadm: add the experimental feature gate PublicKeysECDSA that can be used to create a
cluster with ECDSA certificates from "kubeadm init". Renewal of existing ECDSA certificates is
also supported using "kubeadm alpha certs renew", but not switching between the RSA and
ECDSA algorithms on the fly or during upgrades.
```

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta2/conversion.go:
```diff
@@ -22,7 +22,14 @@ import (
+	}
+
+	out.PublicKeyAlgorithm = &PublicKeyAlgorithm{
+		PublicKeyAlgorithm: in.PublicKeyAlgorithm,
```
v1beta1 and v1beta2 are locked at this point, which means we can only add this as part of v1beta3.
i'm marking https://github.com/kubernetes/kubeadm/issues/1535 as `kind/api-change` so we need to batch include it in the next iteration of the API.

this is a nice change to push the kubeadm config forward.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta2/defaults.go:
```diff
@@ -22,6 +22,7 @@ import (
 
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/runtime"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
```
the public types cannot include the internal types. this is breaking an api-machinery contract.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -17,12 +17,37 @@ limitations under the License.
+	// RSA represents RSA algorithm
+	RSA PublicKeyAlgorithm = iota
+	// ECDSA represents ECDSA algorithm
+	ECDSA
```
can we get away with using elements from the go enumeration directly?
https://golang.org/src/crypto/x509/x509.go

```
type PublicKeyAlgorithm int

const (
	UnknownPublicKeyAlgorithm PublicKeyAlgorithm = iota
	RSA
	DSA
	ECDSA
	Ed25519
)
```

e.g. `x509.ECDSA`

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/types.go:
```diff
@@ -54,6 +79,9 @@ type InitConfiguration struct {
 	CertificateKey string
+
+	// PublicKeyAlgorithm sets the algorithm used for generating encryption keys: RSA or ECDSA.
+	PublicKeyAlgorithm PublicKeyAlgorithm
```
InitConfiguration vs something else is an interesting topic.
do we gain anything if this is part of ClusterConfiguration?

i'm hoping that nobody would like to switch algorithms on a working cluster, but unfortunately people will want to do that on their existing RSA based clusters. this implies CA rotation, so it opens a whole can of worms.

i'm starting to think that this feature needed a KEP (enhancement proposal that clarifies all the details).

rosti replied:
It makes more sense to me to have this field in the ClusterConfiguration. I am too concerned if certificate renewal will work as expected when switching from RSA to ECDSA.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta1/conversion.go:
```diff
@@ -31,6 +31,10 @@ func Convert_kubeadm_InitConfiguration_To_v1beta1_InitConfiguration(in *kubeadm.
 	}
 
+	if in.PublicKeyAlgorithm != kubeadm.RSA {
+		return errors.New("publicKeyAlgorithm field is not supported by v1beta1 config format")
```
as mentioned earlier, old config types should not know about this field and just error out.

neolit123 commented on cmd/kubeadm/app/apis/kubeadm/v1beta2/public_key_algorithm.go:
```diff
@@ -0,0 +1,56 @@
+)
+
+// PublicKeyAlgorithm is a wrapper around kubeadm.PublicKeyAlgorithm which supports
+// correct marshaling to YAML and JSON. In particular, it marshals into strings.
```
instead of using `int` in the internal config (also not use the go enumeration) we might as well use "string" in the internal configuration and this can avoid having this file.

it would still need to convert from the go enumeration to our string enumeration.

rosti replied:
Indeed, we don't need this file. Let's use strings and string consts here.

neolit123 commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -101,7 +101,7 @@ func CreateCACertAndKeyFiles(certSpec *KubeadmCert, cfg *kubeadmapi.InitConfigur
 		return errors.Errorf("this function should only be used for CAs, but cert %s has CA %s", certSpec.Name, certSpec.CAName)
 	}
-	klog.V(1).Infof("creating a new certificate authority for %s", certSpec.Name)
+	klog.V(1).Infof("creating a new certificate authority for %s (public key algo: %s)", certSpec.Name, cfg.PublicKeyAlgorithm)
```
algo -> algorithm

neolit123 reviewed (commented):
@rojkov 
thanks for working on this.

/cc @rosti on the API change.
we need to batch for v1beta3.

rosti reviewed (commented):
Thanks @rojkov !
It's the config representation and possible upgrade issues that bother me here.
We should use simple string consts in the config and currently only in the internal types. The new public field should be added in v1beta3 when we decide to have it.
One possibility to enable it externally is to use an alpha feature gate. We probably need more time discussing it.

fedebongio commented:
/assign @yliaog to take a look at the client-go part

yliaog commented:
for client-go
/lgtm

rojkov commented:
/test pull-kubernetes-e2e-gce

rojkov commented:
Thank you all for quick review! And sorry for delay :)

I reworked the PR so that it uses a feature gate now to set the type of keys as suggested in https://github.com/kubernetes/kubeadm/issues/1535 and removed all the modifications of internal API types. `RSA` and `ECDSA` are just string literals now.

The current logic is that a new cluster created with `--feature-gates=PublicKeysECDSA=true` will use ECDSA upon init and renewal with no opt-out.

Also as `k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm` (`kubeadmapi`) started
to depend on `k8s.io/kubernetes/cmd/kubeadm/app/features` (`features`) the
function `constants.GetDNSVersion()` was moved to `kubeadmapi` as method
`Version()` of `DNSAddOnType`. This was done to avoid the dependency cycle
```
  features -> constants -> kubeadmapi -> features
```

neolit123 commented:
thanks for the update @rojkov 
@kubernetes/sig-cluster-lifecycle-pr-reviews 

the deadline for KEPs is today. if you think this is something we can enable in the API, support upgrades and switching between algorithm types, we'd need a KEP for a future release.


> Does this PR introduce a user-facing change?:
> One can choose ECDSA to generate cluster certificates in kubeadm by specifying the
publicKeyAlgoritm field of the InitConfiguration config.

i think we should add the following note instead:
> kubeadm: add the experimental feature gate PublicKeysECDSA that can be used to create a cluster with ECDSA certificates from "kubeadm init". Renewal of existing ECDSA certificates is also supported using "kubeadm alpha certs renew", but not switching between the RSA and ECDSA algorithms on the fly or during upgrades.

making it slightly more detailed and "official". please amend the details if needed.

rojkov commented:
@neolit123 Thanks! I updated the notes.

There's no chance I can write the KEP today unfortunately, but the feature not urgent for us now.

rojkov commented:
/test pull-kubernetes-e2e-gce

rojkov commented:
/test pull-kubernetes-e2e-gce

neolit123 commented:
/assign @neolit123

neolit123 commented:
/assign @rosti

neolit123 commented:
@rojkov sorry for the delays on reviews, everyone is busy again.

so i think merging this is fine.
/approve
/priority important-longterm

i will have another pass tomorrow and potentially apply the LGTM label too.

neolit123 commented on staging/src/k8s.io/client-go/util/cert/cert.go:
```diff
@@ -39,10 +39,11 @@ const duration365d = time.Hour * 24 * 365
+	Organization       []string
+	AltNames           AltNames
+	Usages             []x509.ExtKeyUsage
+	PublicKeyAlgorithm x509.PublicKeyAlgorithm
```
@rojkov i pinged the #sig-api-machinery channel about this change and it is not approved.
https://kubernetes.slack.com/archives/C0EG7JC6T/p1582187786382200

i guess the alternative is to pass the algorithm type as an argument to functions that need it.

rojkov replied:
Ok! I'll follow @rosti's suggestion to introduce a wrapper struct.

rosti commented on staging/src/k8s.io/client-go/util/cert/cert.go:
```diff
@@ -39,10 +39,11 @@ const duration365d = time.Hour * 24 * 365
+	Organization       []string
+	AltNames           AltNames
+	Usages             []x509.ExtKeyUsage
+	PublicKeyAlgorithm x509.PublicKeyAlgorithm
```
Do we really need this setting here? We can create a kubeadm wrapper struct that embeds this one and adds the `PublicKeyAlgorithm` field.

rosti reviewed (changes requested):
Thanks @rojkov !
I've got a few comments left.

Commit:
kubeadm: allow creating a cluster with ECDSA keys

The selected key type is defined by kubeadm's --feature-gates option:
if it contains PublicKeysECDSA=true then ECDSA keys will be generated
and used.

By default RSA keys are used still.

Signed-off-by: Dmitry Rozhkov <dmitry.rozhkov@linux.intel.com>

rojkov commented:
@neolit123 @rosti Thank you for review! I've updated the PR.

Still I had to touch seemingly unrelated `features_test.go` to break the import cycle. Let me know if it's better to submit a separate PR for that.

neolit123 commented on cmd/kubeadm/app/features/features_test.go:
```diff
@@ -129,7 +129,7 @@ func TestNewFeatureGate(t *testing.T) {
 	var someFeatures = FeatureList{
 		"feature1": {FeatureSpec: featuregate.FeatureSpec{Default: false, PreRelease: featuregate.Beta}},
-		"feature2": {FeatureSpec: featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Alpha}, MinimumVersion: constants.MinimumControlPlaneVersion.WithPreRelease("alpha.1")},
+		"feature2": {FeatureSpec: featuregate.FeatureSpec{Default: true, PreRelease: featuregate.Alpha}, MinimumVersion: version.MustParseSemantic("v1.17.0").WithPreRelease("alpha.1")},
```
i think this is fine, but wonder why it broke related to this change:

> Still I had to touch seemingly unrelated features_test.go to break the import cycle. Let me know if it's better to submit a separate PR for that.

rojkov replied:
This `features` imports `constants` which imports `kubeadmapi` which is going to import `features` in this PR.

neolit123 replied:
ok, understood.

rosti replied:
Actually `GetDNSVersion` should be in the DNS addon package, but that way we get circular dependency through the images package...

neolit123 commented on cmd/kubeadm/app/util/certs/util.go:
```diff
@@ -26,13 +26,16 @@ import (
 
 	certutil "k8s.io/client-go/util/cert"
 	"k8s.io/client-go/util/keyutil"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
```
CertConfig should be defined here, so that the kubeadm API does not have to be imported in the utils.
ultimately it is an utility structure similarly to the wrapped one from client-go.

rojkov replied:
Thanks! Will do.
This should fix the failing typecheck test as well.

rojkov replied:
@neolit123 I had to move `CertConfig` to `pkiutil`, otherwise there would be another import loop (`certs/util` ->  `pkiutil` -> `certs/util`).

Perhaps it makes sense to merge `certs/util` into `pkiutil` at some point.

neolit123 replied:
i think what i meant was "util/pkiutil" and not "util/certs".
it seems like the better location.

neolit123 commented:
this looks fine to me.
@rosti any more comments?

rosti reviewed (approved):
Thanks @rojkov !
/lgtm
